### PR TITLE
[ECO-5127] Implement CHA-RL9

### LINKED ISSUE #131: Implement CHA-RL9 (clarifications for presence wait-for-attach)
https://github.com/ably/specification/pull/232 adds a new spec point to address some of the stuff I couldn’t in #112; implement.

┆Issue is synchronized with this [Jira Story](https://ably.atlassian.net/browse/ECO-5127) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
Resolves #131.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new error case for room state management, enhancing error reporting in the Ably Chat SDK.
  
- **Bug Fixes**
  - Improved clarity in handling room status changes during presence operations, ensuring accurate error reporting.

- **Documentation**
  - Updated documentation for the `FeatureChannel` protocol to clarify behavior during room status changes.

- **Tests**
  - Renamed and adjusted tests to better reflect the expected behavior of room status transitions.

### CODE REVIEW

File: Sources/AblyChat/RoomLifecycleManager.swift [hunk 1]
```diff
@@ -1192,31 +1192,60 @@ internal actor DefaultRoomLifecycleManager<Contributor: RoomLifecycleContributor
...
+            // CHA-RL9a
+            let subscription = onRoomStatusChange(bufferingPolicy: .unbounded)
+            // TODO: what if nil?
+            logger.log(message: "waitToBeAbleToPerformPresenceOperations waiting for status change", level: .debug)
+            #if DEBUG
+                for subscription in statusChangeWaitEventSubscriptions {
+                    subscription.emit(.init())
+                }
+            #endif
+            let nextRoomStatusChange = await (subscription.first { _ in true })!
```
[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift]: Changed it to throw an error and added a TODO linked to an existing GitHub issue for handling task cancellation

File: Sources/AblyChat/RoomLifecycleManager.swift [hunk 2]
```diff
@@ -1192,31 +1192,60 @@ internal actor DefaultRoomLifecycleManager<Contributor: RoomLifecycleContributor
...
+            // CHA-RL9b
+            // TODO: decide what to do if nextRoomStatusChange is nil; I believe that this will happen if the current Task is cancelled. For now, will just treat it as an invalid status change. Handle it properly in https://github.com/ably-labs/ably-chat-swift/issues/29
+            if nextRoomStatusChange?.current != .attached {
+                // CHA-RL9c
+                throw .init(chatError: .roomInInvalidState(cause: nextRoomStatusChange?.current.error))
+            }
         case .attached:
             // CHA-PR3e, CHA-PR11e, CHA-PR6d, CHA-T2d
             break
-        case .detached, .detachedDueToRetryOperation:
```
Comment (maratal) [inline Sources/AblyChat/RoomLifecycleManager.swift:1208]: Where is it gone? No retry anymore?

Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:1208]: The `switch` has changed — it was previously `switch status` (a `Status`, which has multiple statuses for a single `RoomStatus`) but it's now `switch status.roomStatus` (a `RoomStatus`)

**Review by maratal — APPROVED** [review]
Lgtm

**Pull request merged**